File: movil/lib/services/resident_service.dart
import 'dart:convert';
import '../api/api_client.dart';

class ResidentService {
  final ApiClient api;
  ResidentService(this.api);

  Future<List<Map<String, dynamic>>> listResidents({Map<String, dynamic>? params}) async {
    final res = await api.get('/api/residentes/', query: params);
    if (res.statusCode == 200) {
      final data = jsonDecode(res.body);
      if (data is List) return data.cast<Map<String, dynamic>>();
      if (data is Map && data['results'] is List) {
        return (data['results'] as List).cast<Map<String, dynamic>>();
      }
      return [];
    }
    throw Exception('Error ${res.statusCode}');
  }

  Future<Map<String, dynamic>> createResident(Map<String, dynamic> payload) async {
    final res = await api.post('/api/residentes/', body: jsonEncode(payload));
    if (res.statusCode >= 200 && res.statusCode < 300) {
      return api.decodeJson<Map<String, dynamic>>(res);
    }
    throw Exception('Error ${res.statusCode}: ${res.body}');
  }

  Future<Map<String, dynamic>> updateResident(int id, Map<String, dynamic> payload) async {
    final res = await api.patch('/api/residentes/$id/', body: jsonEncode(payload));
    if (res.statusCode >= 200 && res.statusCode < 300) {
      return api.decodeJson<Map<String, dynamic>>(res);
    }
    throw Exception('Error ${res.statusCode}: ${res.body}');
  }
}

File: web_Frontend/src/api/residentService.js
import axios from 'axios';

const API_URL = 'http://localhost:8000/api';

const withAuth = () => {
  const token = localStorage.getItem('token');
  if (token) {
    axios.defaults.headers.common['Authorization'] = `Bearer ${token}`;
  }
};

export const residentService = {
  listFamilies: async (params = {}) => {
    withAuth();
    const res = await axios.get(`${API_URL}/familias/`, { params });
    const data = res.data;
    return Array.isArray(data) ? data : (data.results || []);
  },
  getFamily: async (id) => {
    withAuth();
    const res = await axios.get(`${API_URL}/familias/${id}/`);
    return res.data;
  },
  listResidents: async (params = {}) => {
    withAuth();
    const res = await axios.get(`${API_URL}/residentes/`, { params });
    const data = res.data;
    return Array.isArray(data) ? data : (data.results || []);
  },
  getResident: async (id) => {
    withAuth();
    const res = await axios.get(`${API_URL}/residentes/${id}/`);
    return res.data;
  },
  createResident: async (payload) => {
    withAuth();
    const res = await axios.post(`${API_URL}/residentes/`, payload);
    return res.data;
  },
  updateResident: async (id, payload) => {
    withAuth();
    const res = await axios.patch(`${API_URL}/residentes/${id}/`, payload);
    return res.data;
  },
  createFamily: async (payload) => {
    withAuth();
    const res = await axios.post(`${API_URL}/familias/`, payload);
    return res.data;
  },
  updateFamily: async (id, payload) => {
    withAuth();
    const res = await axios.patch(`${API_URL}/familias/${id}/`, payload);
    return res.data;
  },
  deleteFamily: async (id) => {
    withAuth();
    const res = await axios.delete(`${API_URL}/familias/${id}/`);
    return res.data;
  }
};
